File: chapter_9/src/main/java/com/wzc/chapter_9/MainActivity.java
package com.wzc.chapter_9;

import static com.wzc.chapter_9.MyDynamicReceiver.ACTION_LAUNCH_DYNAMIC;
import static com.wzc.chapter_9.MyStaticReceiver.ACTION_LAUNCH_STATIC;

import android.app.Activity;
import android.content.ComponentName;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.graphics.PixelFormat;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.os.UserHandle;
import android.provider.UserDictionary;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.WindowManager;
import android.os.Process;

public class MainActivity extends Activity {

    private static final String TAG = "MainActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        UserHandle userHandle = Process.myUserHandle();
        Intent intent = new Intent();
        intent.setAction(ACTION_LAUNCH_DYNAMIC);
        String resolvedType = intent.resolveTypeIfNeeded(getContentResolver());
        Uri uri = ContentUris.withAppendedId(UserDictionary.Words.CONTENT_URI, 4);
        uri = ContentUris.appendId(new Uri.Builder().scheme("content").authority(UserDictionary.AUTHORITY).path("words"), 4).build();
        long id = ContentUris.parseId(uri);
        Log.d(TAG, "onCreate: uri = " + uri + ", id = " + id);
    }

    public void openWindow(View view) {
        WindowManager windowManager = (WindowManager) getSystemService(Context.WINDOW_SERVICE);
        View v = getLayoutInflater().inflate(R.layout.window_layout, null);
        v.findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                MainActivity.start(v.getContext());
            }
        });
        WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT, WindowManager.LayoutParams.WRAP_CONTENT,
                0, 0, PixelFormat.TRANSPARENT);
        layoutParams.type = WindowManager.LayoutParams.TYPE_SYSTEM_ALERT;
        layoutParams.gravity = Gravity.CENTER;
        layoutParams.flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                | WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL;
        if (windowManager != null) {
            windowManager.addView(v, layoutParams);
        }
    }

    public static void start(Context context) {
        Intent starter = new Intent(context, MainActivity.class);
        starter.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(starter);
    }

    public void openSecondActivity(View view) {
        Intent intent = new Intent(MainActivity.this, SecondActivity.class);
        MainActivity.this.startActivity(intent);
    }

    public void startService(View view) {
        Intent intent = new Intent(MainActivity.this, MyService.class);
        startService(intent);
    }

    public void stopService(View view) {
        Intent intent = new Intent(MainActivity.this, MyService.class);
        stopService(intent);
    }

    private final ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.d(TAG, "onServiceConnected: ");
            ICompute iCompute = ICompute.Stub.asInterface(service);
            try {
                int result = iCompute.add(1, 1);
                Log.d(TAG, "onServiceConnected: result=" + result + ", processName=" + MyUtils.getProcessName(MainActivity.this, Process.myPid()) + ", threadName=" + Thread.currentThread().getName());
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            Log.d(TAG, "onServiceDisconnected: ");
        }
    };

    public void bindService(View view) {
        Intent intent = new Intent(MainActivity.this, RemoteBindService.class);
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
    }

    public void unbindService(View view) {
        unbindService(serviceConnection);
    }

    private MyDynamicReceiver receiver;
    public void registerReceiver(View view) {
        if (null == receiver) {
            IntentFilter intentFilter = new IntentFilter();
            intentFilter.addAction(ACTION_LAUNCH_DYNAMIC);
            receiver = new MyDynamicReceiver();
            registerReceiver(receiver, intentFilter);
        }
    }

    public void unregisterReceiver(View view) {
        if (null != receiver) {
            unregisterReceiver(receiver);
            receiver = null;
        }
    }

    public void sendBroadcast(View view) {
        Intent intent = new Intent();
        intent.setAction(ACTION_LAUNCH_STATIC);
        sendBroadcast(intent);
    }

    public void openUserDictionary(View view) {
        UserDictionaryActivity.start(this);
    }
}

File: chapter_2_aidl_in_out/src/main/java/com/example/aidl/MainActivity.kt
package com.example.aidl

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import java.util.concurrent.atomic.AtomicBoolean

class MainActivity : Activity() {
    private val serviceBound = AtomicBoolean(false)
    private var iController: IController? = null
    private val serviceConnection = object: ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            Log.d(TAG, "onServiceConnected: currThread=${Thread.currentThread().name}")
            serviceBound.set(true)
            iController = IController.Stub.asInterface(service)
            registerCallback()
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            Log.d(TAG, "onServiceDisconnected: currThread=${Thread.currentThread().name}")
            serviceBound.set(false)
        }
    }

    private val callback = object: ICallback.Stub() {
        override fun onResult(result: Int) {
            Log.d(TAG, "onResult: result=$result, currThread=${Thread.currentThread().name}")
        }
    }

    private val stateIn = State(1)
    private val stateOut = State(1)
    private val stateInOut = State(1)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)
        findViewById(R.id.btn_bind_service).setOnClickListener {
            if (serviceBound.get()) return@setOnClickListener
            val intent = Intent()
            intent.component = ComponentName(this@MainActivity, RemoteService::class.java)
            bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
        }
        findViewById(R.id.btn_unbind_service).setOnClickListener {
            if (serviceBound.get()) {
                unbindService(serviceConnection)
                serviceBound.set(false)
            }
        }
        findViewById(R.id.btn_transIn).setOnClickListener {
            if (serviceBound.get()) {
                Log.d(TAG, "transIn start: stateIn.value=${stateIn.value}")
                try {
                    iController?.transIn(stateIn)
                } catch (e: Exception) {
                }
                Log.d(TAG, "transIn end: stateIn.value=${stateIn.value}")
            }
        }
        findViewById(R.id.btn_transOut).setOnClickListener {
            if (serviceBound.get()) {
                Log.d(TAG, "transOut start: stateOut.value=${stateOut.value}")
                try {
                    iController?.transOut(stateOut)
                } catch (e: Exception) {
                }
                Log.d(TAG, "transOut end: stateOut.value=${stateOut.value}")
            }
        }
        findViewById(R.id.btn_transInOut).setOnClickListener {
            if (serviceBound.get()) {
                Log.d(TAG, "transInOut start: stateInOut.value=${stateInOut.value}")
                try {
                    iController?.transInOut(stateInOut)
                } catch (e: Exception) {
                }
                Log.d(TAG, "transInOut end: stateInOut.value=${stateInOut.value}")
            }
        }
    }

    private fun registerCallback() {
        try {
            Log.d(TAG, "registerCallback: callback=$callback, currThread=${Thread.currentThread().name}")
            iController?.registerCallback(callback)
        } catch (e: Exception) {
            Log.e(TAG, "registerCallback: ", e)
        }
    }

    companion object {
        private const val TAG = "MainActivity"
    }
}